azurerm/resource_arm_private_link_service.go

```diff
@@ -0,0 +1,499 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmPrivateLinkService() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmPrivateLinkServiceCreateUpdate,
+		Read:   resourceArmPrivateLinkServiceRead,
+		Update: resourceArmPrivateLinkServiceCreateUpdate,
+		Delete: resourceArmPrivateLinkServiceDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: aznet.ValidatePrivateLinkServiceName,
+			},
+
+			"location": azure.SchemaLocation(),
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"auto_approval_subscription_ids": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validate.GUID,
+				},
+			},
+
+			"visibility_subscription_ids": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validate.GUID,
+				},
+			},
+
+			// currently not implemented yet, timeline unknown, exact purpose unknown, maybe coming to a future API near you
+			// "fqdns": {
+			// 	Type:     schema.TypeList,
+			// 	Optional: true,
+			// 	Elem: &schema.Schema{
+			// 		Type:         schema.TypeString,
+			// 		ValidateFunc: validate.NoEmptyStrings,
+			// 	},
+			// },
+
+			// Required by the API you can't create the resource without at least one ip configuration
+			// I had to split the schema because if you attempt to change the primary attribute it
+			// succeeds but doesn't change the attribute. Once primary is set it is set forever unless
+			// you destroy the resource and recreate it.
+			"primary_nat_ip_configuration": {
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: aznet.ValidatePrivateLinkServiceName,
+						},
+						"private_ip_address": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validate.IPv4Address,
+						},
+						// Only IPv4 is supported by the API, but I am exposing this
+						// as they will support IPv6 in a future release.
+						"private_ip_address_version": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(network.IPv4),
+							}, false),
+							Default: string(network.IPv4),
+						},
+						"subnet_id": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: azure.ValidateResourceID,
+						},
+					},
+				},
+			},
+
+			"auxillery_nat_ip_configuration": {
+				Type:     schema.TypeList,
```

Comment:
Just realized these should probably be `TypeSet`?

Reply:
Fixed.